### ISSUE TITLE
Function references couldn't be released after `context.release()`

### ISSUE THREAD

**Issue by xtforgame** [laverdet/isolated-vm#112]
node version: v10.11.0
OS version: macOS 10.14.5

The following code is modified from [context-leak.js](https://github.com/laverdet/isolated-vm/blob/master/tests/context-leak.js)

```js
const ivm = require('isolated-vm');

const loop = () => {
  const snapshot = ivm.Isolate.createSnapshot([
    { code: 'function fn(){};' },
  ]);
  const isolate = new ivm.Isolate({ memoryLimit: 32, snapshot });
  const context = isolate.createContextSync();
  const { global } = context;

  global.setSync('global', global.derefInto());
  global.setSync('testFunc', new ivm.Reference(function testFunc() {}));

  context.release();
  isolate.dispose();

  setTimeout(() => {
    loop();
  }, 2);
};

loop();
```

Run `node --inspect test.js` and open `chrome://inspect`, see the `memory` tab

<img width="777" alt="螢幕快照 2019-05-28 上午10 30 38" src="https://user-images.githubusercontent.com/7076799/58446560-b4e91480-8133-11e9-845d-085963cdc77d.png">

Everything looks good!

But if I move `context.release();` and `isolate.dispose();` into the `setTimeout` function, as below:

```js
setTimeout(() => {
  context.release();
  isolate.dispose();
  loop();
}, 2);
```

And run the test for a while, this time, the `testFunc` cannot be released by GC properly

<img width="903" alt="螢幕快照 2019-05-28 上午10 39 30" src="https://user-images.githubusercontent.com/7076799/58446908-270e2900-8135-11e9-862b-dc71d8fdd9f4.png">

Is it a potential memory leak issue, or I missed doing something to release the function reference ?

**Comment by xtforgame** [comment]
OK, I think I have figured out what was going on. That was my bad.

The code works as expect 

```js
const ivm = require('isolated-vm');

let ref = null;

const loop = () => {
  const snapshot = ivm.Isolate.createSnapshot([
    { code: 'function fn(){};' },
  ]);
  const isolate = new ivm.Isolate({ memoryLimit: 32, snapshot });
  const context = isolate.createContextSync();
  const { global } = context;

  ref = new ivm.Reference(function testFunc() {});
  global.setSync('testFunc', ref);

  context.release();
  isolate.dispose();

  setTimeout(() => {
    loop();
  }, 2);
};

loop();
```

I think the root cause is that the `new ivm.Reference(function testFunc() {})` variable left its scope before calling `context.release()`.
While the variable left its scope, it was no longer referenced outside the VM, but it still couldn't be released because the VM is using it inside, after calling `context.release()`, it forms a dangle variable outside the VM.

So if I keep the ivm.Reference in a global variable until calling `context.release()`, it should work

**Comment by laverdet** [comment]
I just had a chance to look at this and you actually did find a bug here. The issue is when using *only* synchronous functions the default nodejs isolate will not release any ivm-related garbage if the last owner of the garbage was a non-default isolate.

I modified your example as follows:

```js
const ivm = require('isolated-vm');

let dummy = new ivm.Isolate;

const loop = () => {
  const isolate = new ivm.Isolate;
  const context = isolate.createContextSync();
  const { global } = context;

  context.global.setSync('testFunc', new ivm.Reference(function testFunc() {}));

  setTimeout(() => {
    dummy.compileScript('1;');
    isolate.dispose();
    loop();
  }, 2);
};

loop();
```

The `dummy.compileScript('1;');` line causes the usual memory management code to run and then the default isolate can cleanup all pending garbage correctly. I suppose the issue went unnoticed for so long because most users are using the promise APIs.

Anyway, this is now fixed in b5429bd77bba07e9daf9434c781ac40928ac742a, thanks for the report!
